File: packages/sdk-typescript/src/idempotency.ts
const MINIMUM_KEY_BYTES = 16;
const MAXIMUM_KEY_BYTES = 255;

// Generate this once per logical mutation and persist it until the operation's
// retry lifecycle is complete. A fresh key represents a new operation.
export const createIdempotencyKey = (
  cryptoProvider: Pick<Crypto, "getRandomValues"> = globalThis.crypto,
): string => {
  if (!cryptoProvider?.getRandomValues) {
    throw new Error("A cryptographically secure random source is required");
  }
  const bytes = cryptoProvider.getRandomValues(new Uint8Array(32));
  return Array.from(bytes, (value) => value.toString(16).padStart(2, "0")).join(
    "",
  );
};

export const validateIdempotencyKey = (value: string): void => {
  const encoded = new TextEncoder().encode(value);
  if (
    encoded.length < MINIMUM_KEY_BYTES ||
    encoded.length > MAXIMUM_KEY_BYTES ||
    !/^[!-~]+$/u.test(value)
  ) {
    throw new Error(
      "FortyOne idempotency key must contain 16 to 255 visible ASCII characters",
    );
  }
};

File: apps/server/sdk/go/idempotency.go
package fortyone

import (
	"crypto/rand"
	"encoding/hex"
	"errors"
)

const (
	minimumIdempotencyKeyBytes = 16
	maximumIdempotencyKeyBytes = 255
)

// NewIdempotencyKey returns a cryptographically random, URL/header-safe key.
// Persist the returned value with the operation until its retry lifecycle is
// complete; generating a fresh key for a retry defeats idempotency.
func NewIdempotencyKey() (string, error) {
	random := make([]byte, 32)
	if _, err := rand.Read(random); err != nil {
		return "", err
	}
	return hex.EncodeToString(random), nil
}

// ValidateIdempotencyKey applies the public API's byte-level header contract.
func ValidateIdempotencyKey(value string) error {
	if len(value) < minimumIdempotencyKeyBytes || len(value) > maximumIdempotencyKeyBytes {
		return errors.New("FortyOne idempotency key must contain 16 to 255 visible ASCII characters")
	}
	for index := 0; index < len(value); index++ {
		if value[index] < 0x21 || value[index] > 0x7e {
			return errors.New("FortyOne idempotency key must contain 16 to 255 visible ASCII characters")
		}
	}
	return nil
}
